File: test-api-pkg/ts/test-docs-workflows.ts
/**
 * Mirrors website/public/sdk/workflows.md against @imbrace/sdk@1.0.4 (npm)
 * — API-key auth.
 *
 * Doc-gap (TS only): the doc shows `client.activepieces.*` for §Activepieces.
 * SDK 1.0.4 only exposes `client.workflows.*`. The Py snippets in the doc
 * already use `client.workflows.*` so no Py rename is needed.
 */
import "dotenv/config"
import { ImbraceClient } from "@imbrace/sdk"

const apiKey         = process.env.IMBRACE_API_KEY
const organizationId = process.env.IMBRACE_ORGANIZATION_ID
const baseUrl        = process.env.IMBRACE_GATEWAY_URL ?? "https://app-gatewayv2.imbrace.co"

if (!apiKey || !organizationId) {
  console.error("Missing IMBRACE_API_KEY or IMBRACE_ORGANIZATION_ID")
  process.exit(1)
}

const client = new ImbraceClient({ apiKey, organizationId, baseUrl, timeout: 30_000 })

let pass = 0, fail = 0, skip = 0
const fails: string[] = []
const docGaps: string[] = []

async function step(label: string, fn: () => Promise<unknown>, expectFail = false) {
  process.stdout.write(`  • ${label} ... `)
  try {
    const t0 = Date.now()
    const r = await fn()
    const dt = Date.now() - t0
    const summary = JSON.stringify(r ?? {}).slice(0, 90)
    if (expectFail) { console.log(`unexpected pass [${dt}ms]: ${summary}`); fail++; fails.push(`${label} → unexpected pass`) }
    else { console.log(`✓ [${dt}ms] ${summary}`); pass++ }
  } catch (err: any) {
    const code = err?.statusCode ?? err?.message ?? "ERR"
    if (expectFail) { console.log(`✓ (expected fail [${code}])`); pass++ }
    else { console.log(`✗ [${code}]`); fail++; fails.push(`${label} → ${code}`) }
  }
}

function skipped(label: string, reason: string) { console.log(`  - ${label}  ⏭ ${reason}`); skip++ }
function section(title: string) { console.log(`\n══ ${title} ══`) }
function note(msg: string) { console.log(`  ℹ ${msg}`); docGaps.push(msg) }

console.log(`\n━━━ DOCS: workflows.md — auth: API KEY (npm @imbrace/sdk@1.0.4) ━━━`)
console.log(`gateway=${baseUrl}\norg=${organizationId}\n`)

const ts = Date.now()
const state: { projectId: string | null; flowId: string | null; runId: string | null; folderId: string | null } =
  { projectId: null, flowId: null, runId: null, folderId: null }

note("doc-gap: workflows.md TS snippets call `client.activepieces.*` everywhere — SDK 1.0.4 only exposes `client.workflows.*`. Py snippets already use `workflows`.")

// ── §1. Flows lifecycle ───────────────────────────────────────────────────

section("§1. Flows (list / get / create / delete)")

await step("workflows.listFlows + capture projectId", async () => {
  const r: any = await client.workflows.listFlows({ limit: 5 })
  state.projectId = r?.data?.[0]?.projectId ?? null
  return { count: r?.data?.length, projectId: state.projectId }
})

if (state.projectId) {
  await step("workflows.createFlow", async () => {
    const f: any = await client.workflows.createFlow({
      displayName: `New Lead Notification ${ts}`,
      projectId: state.projectId!,
    } as any)
    state.flowId = f?.id ?? f?._id ?? null
    return { id: state.flowId }
  })
} else {
  skipped("workflows.createFlow", "no projectId fixture")
}

if (state.flowId) {
  await step("workflows.getFlow", () => (client.workflows as any).getFlow(state.flowId!))
} else {
  skipped("workflows.getFlow", "no flow fixture")
}

// ── §2. Trigger a flow ────────────────────────────────────────────────────

section("§2. Trigger a flow (async + sync)")

if (state.flowId) {
  // Newly-created flow has no trigger configured — triggerFlow may 404.
  // Add the Webhook trigger + publish first so triggerFlow has a target.
  await step("workflows.applyFlowOperation UPDATE_TRIGGER (Webhook)", () =>
    (client.workflows as any).applyFlowOperation(state.flowId!, {
      type: "UPDATE_TRIGGER",
      request: {
        name: "trigger", type: "PIECE_TRIGGER", valid: true,
        displayName: "Webhook",
        settings: {
          pieceName: "@activepieces/piece-webhook", pieceVersion: "0.1.24",
          triggerName: "catch_webhook",
          input: { authType: "none" }, propertySettings: {},
        },
      },
    }),
  )
  await step("workflows.applyFlowOperation LOCK_AND_PUBLISH", () =>
    (client.workflows as any).applyFlowOperation(state.flowId!, { type: "LOCK_AND_PUBLISH", request: {} }),
  )

  await step("workflows.triggerFlow (async)",
    () => (client.workflows as any).triggerFlow(state.flowId!, { contactId: "contact_xxx", event: "lead_qualified" }))
  await step("workflows.triggerFlowSync (expected timeout — no Return Response action)",
    () => (client.workflows as any).triggerFlowSync(state.flowId!, { contactId: "contact_xxx", event: "lead_qualified" }),
    /* expectFail */ true)
} else {
  skipped("workflows.triggerFlow / triggerFlowSync", "no flow fixture")
}

// ── §3. Runs, folders, connections, tables ───────────────────────────────

section("§3. Runs, folders, connections, tables")

if (state.flowId) {
  await step("workflows.listRuns ({ flowId, limit: 20 })", async () => {
    const r: any = await (client.workflows as any).listRuns({ flowId: state.flowId!, limit: 20 })
    state.runId = r?.data?.[0]?.id ?? null
    return { count: r?.data?.length, sampleRunId: state.runId }
  })
  if (state.runId) {
    await step("workflows.getRun(runId)", () => (client.workflows as any).getRun(state.runId!))
  } else {
    skipped("workflows.getRun", "no run fixture (flow not yet triggered)")
  }
} else {
  skipped("workflows.listRuns / getRun", "no flow fixture")
}

await step("workflows.listFolders", async () => {
  const r: any = await (client.workflows as any).listFolders()
  return { count: (r?.data ?? r ?? []).length }
})

if (state.projectId) {
  await step("workflows.createFolder", async () => {
    const f: any = await (client.workflows as any).createFolder({
      displayName: `CRM Automations ${ts}`, projectId: state.projectId!,
    })
    state.folderId = f?.id ?? f?._id ?? null
    return { id: state.folderId }
  })
} else {
  skipped("workflows.createFolder", "no projectId fixture")
}

await step("workflows.listConnections", async () => {
  const r: any = await (client.workflows as any).listConnections()
  return { count: (r?.data ?? r ?? []).length }
})

skipped("workflows.upsertConnection",
  "destructive — would create a Slack connection in the org. Doc payload uses `xoxb-xxx` placeholder")

await step("workflows.listTables", async () => {
  const r: any = await (client.workflows as any).listTables()
  return { count: (r?.data ?? r ?? []).length }
})

skipped("workflows.listRecords", "needs a real table_id from listTables — fixture-dependent")

// ── §4. Channel automation ────────────────────────────────────────────────

section("§4. Channel automation")

await step("workflows.listChannelAutomation", async () => {
  const r: any = await (client.workflows as any).listChannelAutomation()
  return { count: (r?.data ?? r ?? []).length }
})
await step("workflows.listChannelAutomation ({ channelType: 'whatsapp' })", async () => {
  const r: any = await (client.workflows as any).listChannelAutomation({ channelType: "whatsapp" })
  return { count: (r?.data ?? r ?? []).length }
})

// ── Cleanup ───────────────────────────────────────────────────────────────

section("cleanup")
if (state.flowId) {
  await step("workflows.deleteFlow (cleanup)", () => (client.workflows as any).deleteFlow(state.flowId!))
}
// Folder cleanup — only if SDK exposes deleteFolder
if (state.folderId) {
  await step("workflows.deleteFolder (cleanup; best-effort)",
    () => (client.workflows as any).deleteFolder?.(state.folderId!) ?? Promise.resolve(),
  )
}

console.log(`\n━━━ Summary (workflows / API key) ━━━`)
console.log(`pass=${pass}  fail=${fail}  skip=${skip}`)
if (fails.length) { console.log("Failures:"); fails.forEach(f => console.log(`  - ${f}`)) }
if (docGaps.length) { console.log("Doc gaps to fix:"); docGaps.forEach(g => console.log(`  - ${g}`)) }
process.exit(fail > 0 ? 1 : 0)

File: test-api-pkg/py/test_docs_workflows.py
"""Mirrors website/public/sdk/workflows.md against `imbrace==1.0.4` (PyPI)
— API-key auth.

Note: workflows.md Py snippets already use `client.workflows.*`, so there's
no rename gap on the Py side (TS only).
"""
from __future__ import annotations
import json
import os
import sys
import time

from dotenv import load_dotenv
load_dotenv()

from imbrace import ImbraceClient

API_KEY = os.environ.get("IMBRACE_API_KEY")
ORG_ID  = os.environ.get("IMBRACE_ORGANIZATION_ID")
GATEWAY = os.environ.get("IMBRACE_GATEWAY_URL", "https://app-gatewayv2.imbrace.co")

if not API_KEY or not ORG_ID:
    print("Missing IMBRACE_API_KEY or IMBRACE_ORGANIZATION_ID")
    sys.exit(1)

client = ImbraceClient(api_key=API_KEY, organization_id=ORG_ID, gateway=GATEWAY, timeout=30)

passed = failed = skipped_n = 0
failures: list[str] = []
doc_gaps: list[str] = []


def step(label, fn, expect_fail=False):
    global passed, failed
    sys.stdout.write(f"  - {label} ... "); sys.stdout.flush()
    try:
        t0 = time.time()
        r = fn()
        dt = int((time.time() - t0) * 1000)
        summary = json.dumps(r, default=str)[:90] if r is not None else ""
        if expect_fail:
            print(f"unexpected pass [{dt}ms] {summary}")
            failed += 1; failures.append(f"{label} -> unexpected pass")
        else:
            print(f"OK [{dt}ms] {summary}"); passed += 1
    except Exception as e:
        code = str(e)[:80]
        if expect_fail:
            print(f"OK (expected fail [{code}])"); passed += 1
        else:
            print(f"FAIL [{code}]"); failed += 1
            failures.append(f"{label} -> {code}")


def skip(label, reason):
    global skipped_n
    print(f"  - {label}  SKIP: {reason}"); skipped_n += 1


def section(title): print(f"\n== {title} ==")
def note(msg): print(f"  i {msg}"); doc_gaps.append(msg)


print(f"\n=== DOCS: workflows.md - auth: API KEY (PyPI imbrace==1.0.4) ===")
print(f"gateway={GATEWAY}\norg={ORG_ID}\n")

ts = int(time.time() * 1000)
state = {"projectId": None, "flowId": None, "runId": None, "folderId": None}

note("note: workflows.md Py snippets correctly use `client.workflows.*` (no rename needed). TS doc-gap is `client.activepieces.*` -> `client.workflows.*`.")


section("§1. Flows (list / get / create / delete)")


def _list_flows():
    r = client.workflows.list_flows(limit=5)
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    state["projectId"] = data[0].get("projectId") if data else None
    return {"count": len(data), "projectId": state["projectId"]}


step("workflows.list_flows + capture projectId", _list_flows)


def _create_flow():
    f = client.workflows.create_flow(display_name=f"New Lead Notification {ts}", project_id=state["projectId"])
    state["flowId"] = f.get("id") if isinstance(f, dict) else getattr(f, "id", None)
    return {"id": state["flowId"]}


if state["projectId"]:
    step("workflows.create_flow", _create_flow)
else:
    skip("workflows.create_flow", "no projectId fixture")

if state["flowId"]:
    step("workflows.get_flow", lambda: client.workflows.get_flow(state["flowId"]))
else:
    skip("workflows.get_flow", "no flow fixture")


section("§2. Trigger a flow (async + sync)")
if state["flowId"]:
    step("workflows.apply_flow_operation UPDATE_TRIGGER (Webhook)",
         lambda: client.workflows.apply_flow_operation(state["flowId"], {
             "type": "UPDATE_TRIGGER",
             "request": {
                 "name": "trigger", "type": "PIECE_TRIGGER", "valid": True,
                 "displayName": "Webhook",
                 "settings": {
                     "pieceName": "@activepieces/piece-webhook", "pieceVersion": "0.1.24",
                     "triggerName": "catch_webhook",
                     "input": {"authType": "none"}, "propertySettings": {},
                 },
             },
         }))
    step("workflows.apply_flow_operation LOCK_AND_PUBLISH",
         lambda: client.workflows.apply_flow_operation(state["flowId"], {"type": "LOCK_AND_PUBLISH", "request": {}}))
    step("workflows.trigger_flow (async)",
         lambda: client.workflows.trigger_flow(state["flowId"], {"contactId": "contact_xxx", "event": "lead_qualified"}))
    step("workflows.trigger_flow_sync (expected timeout — no Return Response action)",
         lambda: client.workflows.trigger_flow_sync(state["flowId"], {"contactId": "contact_xxx", "event": "lead_qualified"}),
         expect_fail=True)
else:
    skip("trigger_flow / trigger_flow_sync", "no flow fixture")


section("§3. Runs, folders, connections, tables")
if state["flowId"]:
    def _list_runs():
        r = client.workflows.list_runs(flow_id=state["flowId"], limit=20)
        data = r.get("data", []) if isinstance(r, dict) else (r or [])
        state["runId"] = data[0].get("id") if data else None
        return {"count": len(data), "sampleRunId": state["runId"]}
    step("workflows.list_runs ({ flow_id, limit: 20 })", _list_runs)
    if state["runId"]:
        step("workflows.get_run(run_id)", lambda: client.workflows.get_run(state["runId"]))
    else:
        skip("workflows.get_run", "no run fixture")
else:
    skip("workflows.list_runs / get_run", "no flow fixture")


def _list_folders():
    r = client.workflows.list_folders()
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    return {"count": len(data)}


step("workflows.list_folders", _list_folders)

if state["projectId"]:
    def _create_folder():
        f = client.workflows.create_folder(display_name=f"CRM Automations {ts}", project_id=state["projectId"])
        state["folderId"] = f.get("id") if isinstance(f, dict) else getattr(f, "id", None)
        return {"id": state["folderId"]}
    step("workflows.create_folder", _create_folder)
else:
    skip("workflows.create_folder", "no projectId fixture")


def _list_connections():
    r = client.workflows.list_connections()
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    return {"count": len(data)}


step("workflows.list_connections", _list_connections)
skip("workflows.upsert_connection",
     "destructive — would create a Slack connection in the org. Doc payload uses `xoxb-xxx` placeholder")


def _list_tables():
    r = client.workflows.list_tables()
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    return {"count": len(data)}


step("workflows.list_tables", _list_tables)
skip("workflows.list_records", "needs a real table_id from list_tables — fixture-dependent")


section("§4. Channel automation")


def _list_ca():
    r = client.workflows.list_channel_automation()
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    return {"count": len(data)}


step("workflows.list_channel_automation", _list_ca)


def _list_ca_wa():
    r = client.workflows.list_channel_automation(channel_type="whatsapp")
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    return {"count": len(data)}


step("workflows.list_channel_automation (channel_type='whatsapp')", _list_ca_wa)


section("cleanup")
if state["flowId"]:
    step("workflows.delete_flow (cleanup)", lambda: client.workflows.delete_flow(state["flowId"]))
if state["folderId"] and hasattr(client.workflows, "delete_folder"):
    step("workflows.delete_folder (cleanup; best-effort)",
         lambda: client.workflows.delete_folder(state["folderId"]))


print(f"\n=== Summary (workflows / API key) ===")
print(f"pass={passed}  fail={failed}  skip={skipped_n}")
if failures:
    print("Failures:")
    for f in failures: print(f"  - {f}")
if doc_gaps:
    print("Doc gaps:")
    for g in doc_gaps: print(f"  - {g}")
sys.exit(1 if failed > 0 else 0)
